src/lib/ptree/worker.lua

```diff
@@ -0,0 +1,118 @@
+-- Use of this source code is governed by the Apache 2.0 license; see COPYING.
+
+module(...,package.seeall)
+
+local S            = require("syscall")
+local engine       = require("core.app")
+local app_graph    = require("core.config")
+local counter      = require("core.counter")
+local histogram    = require('core.histogram')
+local lib          = require('core.lib')
+local timer        = require('core.timer')
+local channel      = require("lib.ptree.channel")
+local action_codec = require("lib.ptree.action_codec")
+local alarm_codec  = require("lib.ptree.alarm_codec")
+
+local Worker = {}
+
+local worker_config_spec = {
+   duration = {},
+   measure_latency = {default=true},
+   no_report = {default=false},
+   report = {default={showapps=true,showlinks=true}},
+   Hz = {default=1000},
+}
+
+function new_worker (conf)
+   local conf = lib.parse(conf, worker_config_spec)
+   local ret = setmetatable({}, {__index=Worker})
+   ret.period = 1/conf.Hz
+   ret.duration = conf.duration or 1/0
+   ret.no_report = conf.no_report
+   ret.channel = channel.create('config-worker-channel', 1e6)
+   ret.alarms_channel = alarm_codec.get_channel()
+   ret.pending_actions = {}
+
+   ret.breathe = engine.breathe
+   if conf.measure_latency then
+      local latency = histogram.create('engine/latency.histogram', 1e-6, 1e0)
+      ret.breathe = latency:wrap_thunk(ret.breathe, engine.now)
+   end
+   return ret
+end
+
+function Worker:shutdown()
+   -- This will call stop() on all apps.
+   engine.configure(app_graph.new())
+
+   -- Now we can exit.
+   S.exit(0)
```

Comment:
Why not `main.exit(0)`?

Reply:
I do not know.  This comes from `follower.lua`.  I guess you mean `os.exit(0)` btw.  Can fix I guess :)

Reply:
Actually, `main.exit()` does exist but currently simply calls out to `os.exit()`. I think the idea was to be prepared for potential cleanup code in the engine?